### PR TITLE
Fix Issue 276: advanced search add/remove without exposed operator

### PR DESCRIPTION
**Pull request by patdunlavey**
See https://github.com/esmero/format_strawberryfield/issues/276

The wrapper fieldset is needed for the add/remove mechanism to work. This PR ensures that it will be there, whether or not it started out that way.

### CODE REVIEW

**Comment by DiegoPino** [comment]
@patdunlavey interesting approach. This feels good. Will test with your view and ours and will get back to you tomorrow AM. Thanks so much for putting this work in.

File: modules/format_strawberryfield_views/src/Plugin/views/filter/AdvancedSearchApiFulltext.php [hunk 1]
```diff
@@ -578,15 +578,18 @@ public function buildExposedForm(&$form, FormStateInterface $form_state) {
         '#default_value' => $form_state->getValue($searched_fields_identifier),
       ];
 
-      if (isset($form[$this->options['id'] . '_wrapper'])) {
-        $form[$this->options['id'] . '_wrapper'][$searched_fields_identifier]
-          = $newelements[$searched_fields_identifier];
+      if (empty($form[$this->options['id'] . '_wrapper'])) {
+        $form[$this->options['id'] . '_wrapper'] =
+          [
```
Comment (DiegoPino) [inline modules/format_strawberryfield_views/src/Plugin/views/filter/AdvancedSearchApiFulltext.php:583]: @patdunlavey small thing, can you move the `[` to the previous line? DCS

File: modules/format_strawberryfield_views/src/Plugin/views/filter/AdvancedSearchApiFulltext.php [hunk 2]
```diff
@@ -578,15 +578,18 @@ public function buildExposedForm(&$form, FormStateInterface $form_state) {
...
+            "#type" => "fieldset",
```
Comment (DiegoPino) [inline modules/format_strawberryfield_views/src/Plugin/views/filter/AdvancedSearchApiFulltext.php:584]: @patdunlavey this here should better come from the original Method, in case they decide to change how the wrapper is build (at the extended Filter class)

protected function buildValueWrapper(&$form, $wrapper_identifier) {
    // If both the field and the operator are exposed, this will end up being
    // called twice. We don't want to wipe out what's already there, so if it
    // exists already, do nothing.
    if (!isset($form[$wrapper_identifier])) {
      $form[$wrapper_identifier] = [
        '#type' => 'fieldset',
      ];

      $exposed_info = $this->exposedInfo();
      if (!empty($exposed_info['label'])) {
        $form[$wrapper_identifier]['#title'] = $exposed_info['label'];
      }
      if (!empty($exposed_info['description'])) {
        $form[$wrapper_identifier]['#description'] = $exposed_info['description'];
      }
    }
  }


So basically
$wrapper = $form[$this->options['id'] . '_wrapper'];
$this->buildValueWrapper($form, $wrapper);

That also makes the code slimmer. What do you think?

Comment (patdunlavey) [reply modules/format_strawberryfield_views/src/Plugin/views/filter/AdvancedSearchApiFulltext.php]: Does this fix it?

File: modules/format_strawberryfield_views/src/Plugin/views/filter/AdvancedSearchApiFulltext.php (hunk 1 above)
Comment (patdunlavey) [reply modules/format_strawberryfield_views/src/Plugin/views/filter/AdvancedSearchApiFulltext.php]: This line is removed by using the buildValueWrapper method (see next)

File: modules/format_strawberryfield_views/src/Plugin/views/filter/AdvancedSearchApiFulltext.php [hunk 3]
```diff
@@ -578,18 +578,13 @@ public function buildExposedForm(&$form, FormStateInterface $form_state) {
...
       ];
 
-      if (empty($form[$this->options['id'] . '_wrapper'])) {
-        $form[$this->options['id'] . '_wrapper'] =
-          [
-            "#type" => "fieldset",
-            "#title" => t('Advanced Fulltext search'),
-            $this->options['id'] => $form[$this->options['id']],
-           ];
+      if (empty($form[$this->options['id'] . '_wrapper']) && !empty($form[$this->options['id']])) {
```
Comment (DiegoPino) [inline modules/format_strawberryfield_views/src/Plugin/views/filter/AdvancedSearchApiFulltext.php:581]: Almost. I mean it does work, but if you check the method you don't need too check if the wrapper is already there or not, the  $this->buildValueWrapper() does it already for you.

Comment (patdunlavey) [reply modules/format_strawberryfield_views/src/Plugin/views/filter/AdvancedSearchApiFulltext.php:581]: I put in the check because in the next line we make this assignment:
```
$form[$this->options['id'] . '_wrapper'][$this->options['id']] = $form[$this->options['id']];
```
If the form has already had the wrapper provided (e.g. if you elect to have exposed the operator), then `$form[$this->options['id']]` will be empty and would cause an error.

Comment (patdunlavey) [reply modules/format_strawberryfield_views/src/Plugin/views/filter/AdvancedSearchApiFulltext.php:581]: Ok, I realize the mistake in my argument. Can we always assume the absence of the wrapper element guarantees that `$form[$this->options['id']]` exists? `buildValueWrapper` does not, by itself, guarantee it.

Comment (patdunlavey) [reply modules/format_strawberryfield_views/src/Plugin/views/filter/AdvancedSearchApiFulltext.php]: Ok - removed.

**Review by DiegoPino — APPROVED** [review]
Hi, thanks a lot. Not exactly what I meant but it does work and for the sake of moving forward I I approve and will merging. I will come back to this after that. Thanks so much for your work on this

**Comment by patdunlavey** [comment]
Thanks very much. Sorry for the unresolved confusion! Sometimes, you just gotta spell it out!
